subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_uart.c

```diff
@@ -0,0 +1,303 @@
+	 */
+	USART_HOST_CMD_RX_BAD,
+
+	/*
+	 * Receiving data overrun bytes
+	 * If extra bytes are received after current_state is in complete,
+	 * host is sending extra bytes which indicates data overrun.
+	 */
+	USART_HOST_CMD_RX_OVERRUN,
+};
+
+struct ec_host_cmd_uart_ctx {
+	/* uart device instance */
+	const struct device *uart_dev;
+	/* Context for read and tx operations */
+	struct ec_host_cmd_rx_ctx *rx_ctx;
+	struct ec_host_cmd_tx_buf *tx;
+	struct k_work_delayable timeout_work;
+	enum uart_host_command_state state;
+	uint32_t expected_len;
+};
+
+static int request_expected_size(const struct ec_host_cmd_request_header *r)
+{
+	/* Check host request version */
+	if (r->prtcl_ver != 3) {
+		return 0;
+	}
+
+	/* Reserved byte should be 0 */
+	if (r->reserved) {
+		return 0;
+	}
+
+	return sizeof(*r) + r->data_len;
+}
+
+#define EC_HOST_CMD_UART_DEFINE(_name)                                                             \
+	static struct ec_host_cmd_uart_ctx _name##_hc_uart;                                        \
+	struct ec_host_cmd_backend _name = {                                                       \
+		.api = &ec_host_cmd_api,                                                           \
+		.ctx = (struct ec_host_cmd_uart_ctx *)&_name##_hc_uart,                            \
+	}
+
+/* Waiting time in microseconds to detect overrun */
+#define OVERRUN_TIMEOUT	     300
+/* Timeout after receiving first byte */
+#define USART_REQ_RX_TIMEOUT K_MSEC(150)
+
+static void rx_timeout(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct ec_host_cmd_uart_ctx *hc_uart =
+		CONTAINER_OF(dwork, struct ec_host_cmd_uart_ctx, timeout_work);
+	int res;
+
+	LOG_ERR("USART HOST CMD ERROR: ");
+	switch (hc_uart->state) {
+	case USART_HOST_CMD_RECEIVING:
+		/* If state is receiving then timeout was hit due to underrun */
+		LOG_ERR("Request underrun detected.\n");
+		break;
+	case USART_HOST_CMD_RX_OVERRUN:
+		/* If state is rx_overrun then timeout was hit because
+		 * process request was cancelled and extra rx bytes were
+		 * dropped
+		 */
+		LOG_ERR("Request overrun detected.\n");
+		break;
+	case USART_HOST_CMD_RX_BAD:
+		/* If state is rx_bad then packet header was bad and process
+		 * request was cancelled to drop all incoming bytes.
+		 */
+		LOG_ERR("Bad packet header detected.\n");
+		break;
+	default:
+		LOG_ERR("Request timeout mishandled\n");
+	}
+
+	/* TODO: Workaround - reprogram DMA not to lose data at the end of buffer.
+	 * Remove it once the circular buffer is supported.
+	 */
+	res = uart_rx_disable(hc_uart->uart_dev);
+	res = uart_rx_enable(hc_uart->uart_dev, hc_uart->rx_ctx->buf,
+			     CONFIG_EC_HOST_CMD_HANDLER_RX_BUFFER, OVERRUN_TIMEOUT);
+
+	hc_uart->state = USART_HOST_CMD_READY_TO_RX;
+}
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct ec_host_cmd_uart_ctx *hc_uart = (struct ec_host_cmd_uart_ctx *)user_data;
```

Comment:
Drop the `(struct ec_host_cmd_uart_ctx *)` casting.

Reply:
Done.